File: Codility/Fibonacci.py
"""
Fib[0] = 0
Fib[1] = 1
Fib[n] = Fib[n-1] + Fib[n-2]
"""
from cmath import exp

import pytest


# Method 1: recursion TLE
def Fibonacci_recursion(n: int) -> int:
    if n == 1:
        return 1
    if n == 0:
        return 0
    res = Fibonacci_recursion(n - 1) + Fibonacci_recursion(n - 2)
    return res


# Method 2: dp
def Fibonacci_dp(n: int) -> int:
    fib = [-1] * (n + 1)
    fib[0], fib[1] = 0, 1

    def recursion(m):
        if m == 1:
            return fib[1]
        if m == 0:
            return fib[0]
        if fib[m] != -1:
            return fib[m]
        fib[m] = recursion(m - 1) + recursion(m - 2)
        return fib[m]

    return recursion(n)


test_case1 = {
    (0, 0),
    (1, 1),
    (2, 1),
    (3, 2),
    (4, 3),
    (5, 5),
    (6, 8),
    (20, 6765),
}


@pytest.mark.parametrize("n, expect", test_case1)
def test_Fibonacci_recursion(n: int, expect: int) -> None:
    assert Fibonacci_recursion(n) == expect


test_case2 = {
    (100, 354224848179261915075),
    (50, 12586269025),
    (30, 832040),
    (20, 6765),
    (3, 2),
    (4, 3),
    (5, 5),
    (6, 8),
}


@pytest.mark.parametrize("n, expect", test_case2)
def test_Fibonacci_dp(n: int, expect: int) -> None:
    assert Fibonacci_dp(n) == expect

File: Codility/Fibonacci.go
package main

import "fmt"

func Fib(x int) int {
	fib := make([]int, x+1)
	fmt.Println(fib, x)

	return recursion(x, fib)
}

func recursion(x int, fib []int) int {
	if fib[x] != 0 {
		return fib[x]
	}
	fib[0] = 0
	fib[1] = 1
	if x > 1 {
		fib[x] = recursion(x-1, fib) + recursion(x-2, fib)
	}
	return fib[x]
}

func main() {
	fmt.Println(Fib(50))
}
